scio-elasticsearch/es8/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,755 @@
+        }
+
+        return bulkItemResponse;
+      };
+    }
+
+    private static ProcessFunction<BulkRequest, BulkResponse> retry(
+        final ProcessFunction<BulkRequest, BulkResponse> requestFn,
+        final FluentBackoff backoffConfig) {
+      return chunk -> {
+        final BackOff backoff = backoffConfig.backoff();
+        int attempt = 0;
+        BulkResponse response = null;
+        Exception exception = null;
+
+        while (response == null && BackOffUtils.next(Sleeper.DEFAULT, backoff)) {
+          LOG.warn(String.format(RETRY_ATTEMPT_LOG, ++attempt));
+          try {
+            response = requestFn.apply(chunk);
+            exception = null;
+          } catch (Exception e) {
+            exception = e;
+          }
+        }
+
+        if (exception != null) {
+          throw new Exception(String.format(RETRY_FAILED_LOG, attempt), exception);
+        }
+
+        return response;
+      };
+    }
+
+    private static class ClientSupplier implements Supplier<ElasticsearchClient>, Closeable,
+        Serializable {
+
+      private final AtomicReference<RestClientTransport> transport = new AtomicReference<>();
+      private final HttpHost[] nodes;
+      private final UsernamePasswordCredentials credentials;
+
+      public ClientSupplier(final HttpHost[] nodes) {
+        this(nodes, null);
+      }
+
+      public ClientSupplier(final HttpHost[] nodes, UsernamePasswordCredentials credentials) {
+        this.nodes = nodes;
+        this.credentials = credentials;
+      }
+
+      @Override
+      public ElasticsearchClient get() {
+        return new ElasticsearchClient(
+            transport.updateAndGet(c -> c == null ? createTransport() : c));
+
+      }
+
+      @Override
+      public void close() throws IOException {
+        transport.getAndSet(null).close();
+      }
+
+      private RestClientTransport createTransport() {
+        final RestClientBuilder builder = RestClient.builder(nodes);
+        if (credentials != null) {
+          final CredentialsProvider provider = new BasicCredentialsProvider();
+          provider.setCredentials(AuthScope.ANY, credentials);
+          builder.setHttpClientConfigCallback(
+              asyncBuilder -> asyncBuilder.setDefaultCredentialsProvider(provider));
+        }
+        final RestClient restClient = builder.build();
+        return new RestClientTransport(restClient, new JacksonJsonpMapper());
+      }
+    }
+
+    private static ThrowingConsumer<BulkExecutionException> defaultErrorHandler() {
+      return throwable -> {
+        throw throwable;
+      };
+    }
+
+    /**
+     * An exception that puts information about the failures in the bulk execution.
+     */
+    public static class BulkExecutionException extends IOException {
+
+      private final Iterable<ErrorCause> failures;
+
+      BulkExecutionException(BulkResponse bulkResponse) {
+        // super(bulkResponse.items()); TODO
```

Comment:
This does not extend `Throwable` anymore.
Should we at least provide a custom message ?

Reply:
hm, it looks like it still has a [buildFailureMessage](https://github.com/elastic/elasticsearch/blob/v8.1.2/server/src/main/java/org/elasticsearch/action/bulk/BulkResponse.java#L98) method we can use?

Reply:
This isn't available on the Client Java API: [BulkResponse](https://github.com/elastic/elasticsearch-java/blob/main/java-client/src/main/java/co/elastic/clients/elasticsearch/core/BulkResponse.java).

Reply:
created our own `buildFailureMessage` on the same model